ebpf_ffi/cbpf.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ebpf_ffi/cbpf.h"
+
+#include <linux/bpf.h>
+#include <linux/filter.h>
+#include <linux/if_ether.h>
+#include <linux/kernel.h>
+#include <netpacket/packet.h>
+
+namespace cbpf_ffi {
+
+// This constant was determined arbitrarily, the number of 0's has incremented
+// when the size was no longer enough for the verifier logs.
+constexpr size_t kLogBuffSize = 100000000;
+}  // namespace cbpf_ffi
+
+void load_cbpf_program(void *prog_buff, std::string *error, int *socks) {
+  if (socketpair(AF_UNIX, SOCK_DGRAM, 0, socks) != 0) {
```

Comment:
note that on failure socketpair will not modify `socks` https://man7.org/linux/man-pages/man2/socketpair.2.html#:~:text=On%20Linux%20(and%20other%20systems)%2C%20socketpair()%20does%20not%20modify%20sv%20on%0A%20%20%20%20%20%20%20failure.%20%20A%20requirement%20standardizing%20this%20behavior%20was%20added%20in%0A%20%20%20%20%20%20%20POSIX.1%2D2008%20TC2.

Reply:
Ok